design/thin-edge-core.md

```diff
@@ -0,0 +1,278 @@
+from a catalog of generic and use-case specific components assembled
+on top of a framework that enables connectivity and interoperability.__
+
+With IIoT, there is a fantastic opportunity for innovative use-cases and business models
+that are more reactive, flexible, and efficient.
+However, the challenges are many and each require a different expertise.
+A typical IoT application has to address:
+
+* __sophisticated business-cases__, with applications involving several vendors, actors and users
+  and combining connectivity, device management, telemetry, monitoring, analytics;
+* __very diverse integration needs__ over operating systems, protocols, data sources, processing tools,
+  actuators, cloud end-points, legacy systems;
+* __constraint environments__ with low-resources devices, high exposure to security risks,
+  prohibitive manual operations, restricted upgrades;
+* __business-critical or even safety-critical requirements__ with large fleet of devices
+  operating in hazardous contexts and requesting long-term support.
+  
+Furthermore, the forces behind these requirements are pushing in opposite directions.
+
+* On one side, to address the diversity of use-cases and integration requirements,
+  one needs a large open set of features that can be easily extended and combined.
+  [NodeRed](https://nodered.org/) is the archetype of such tools facilitating connections of components.
+* On the other side, to meet resource and security constraints,
+  one needs a minimal system specifically built for a task with cherry-picked components.
+  Taken to the extreme, this leads to the idea of [unikernel](https://en.wikipedia.org/wiki/Unikernel)
+  with sealed, single-purpose software images.
+  
+What makes thin-edge unique is its approach to reconcile these two poles
+with two levels of building blocks that combine along an IIoT specific API.
+* To maximize flexibility and interoperability,
+  thin-edge provides the tools to build an agent from independent executables
+  that interact over MQTT and HTTP using JSON messages.
+* To minimize resources and vulnerabilities,
+  thin-edge features fine-grain Rust components
+  as well as the rules to combine them into hardened executables.
+* To ease incremental developments,
+  thin-edge give the developers the freedom to combine both MQTT and Rust-based components
+  into purpose-specific agents.
+* To make feasible the interoperability of components provided by different tiers,
+  thin-edge comes with an extensible data model for IIoT -
+  cloud connectivity, telemetry data, device management, child devices, ...
+
+## Design Principles
+
+* Thin-edge provides the tools to develop an IIoT agent by assembling components
+  that have been implemented independently,
+  possibly by different vendors and using different programming languages.
+* There are two levels of components:
+  * MQTT-based components that are processes running on the local network
```

Comment:
```suggestion
  * MQTT-based components that are processes running thin-edge devices themselves or on the local network
```

Reply:
Doesn't the local network include the device itself?

Reply:
Made explicit